dialogue.go

```diff
@@ -79,20 +79,23 @@ func (d *Dialogue) MarshalTo(b []byte) error {
 	}
 
 	if field := d.SingleAsn1Type; field != nil {
-		if err := field.MarshalTo(b[offset : offset+field.MarshalLen()]); err != nil {
-			return err
+		if field := d.DialoguePDU; field != nil {
```

Comment:
A little confusing here
* same variable name isn't generally a good idea for readability
* the assigned variable is not used within the scope
* rather than making the nest deeper, I prefer returning earlier

So I'd suggest this;

```go
// MarshalTo puts the byte sequence in the byte array given as b.
func (d *Dialogue) MarshalTo(b []byte) error {
	if len(b) < 4 {
		return io.ErrUnexpectedEOF
	}
	b[0] = uint8(d.Tag)
	b[1] = d.Length
	b[2] = uint8(d.ExternalTag)
	b[3] = d.ExternalLength

	var offset = 4
	if field := d.ObjectIdentifier; field != nil {
		if err := field.MarshalTo(b[offset : offset+field.MarshalLen()]); err != nil {
			return err
		}
		offset += field.MarshalLen()
	}

	if d.SingleAsn1Type == nil {
		copy(b[offset:], d.Payload)
		return nil
	}

	if field := d.DialoguePDU; field != nil {
		d.SingleAsn1Type.Value = make([]byte, field.MarshalLen())
		if err := field.MarshalTo(d.SingleAsn1Type.Value); err != nil {
			return err
		}
	}

	d.SingleAsn1Type.SetLength()
	if err := d.SingleAsn1Type.MarshalTo(b[offset : offset+d.SingleAsn1Type.MarshalLen()]); err != nil {
		return err
	}
	offset += d.SingleAsn1Type.MarshalLen()

	copy(b[offset:], d.Payload)

	return nil
}
```